dsp/dsp.cpp

```diff
@@ -227,6 +232,27 @@ void tanh_(Eigen::MatrixXf &x) {
   }
 }
 
+void hard_tanh_(Eigen::MatrixXf& x, const long i_start, const long i_end,
+    const long j_start, const long j_end) {
+    for (long j = j_start; j < j_end; j++)
+        for (long i = i_start; i < i_end; i++)
+            x(i, j) = hard_tanh_(x(i, j));
```

Comment:
I wonder if it would perform better if you used unaryExpr instead? Like so:
  x = x.unaryExpr([](float in) {return hard_tanh(in);});

Seems to do better according to: https://godbolt.org/z/zfePz8MTa

Reply:
Well, you don't want to update the entire matrix, so some x.middleCols() are needed

Reply:
> I wonder if it would perform better if you used unaryExpr instead? Like so:
> x = x.unaryExpr([](float in) {return hard_tanh(in);});

Agreed - I just copy/pasted a duplicate of the tanh function structure, but didn't worry about it for now as that code path isn't used by WaveNet - it just uses the full matrix overload and I've already optimized those (for both tanh and hardtanh).

My plan is to refactor the activation code in general soon, but I wanted to get the hardtanh in there asap so we can start think about targeting it with training models.

Reply:
Btw, I tried using unaryExpr(), and if I recall it performed slightly worse than just rolling over the data directly.

That Compiler Explorer tool you linked to is pretty cool.